### PR TITLE
[Merged by Bors] - refactor: Closure predicate for `ClosureOperator`

### PR DESCRIPTION
**Pull request by YaelDillies**
This PR is the result of two years and a half of pondering.

The only way we ever define closure operators is by having a pre-existing "closedness" predicate and defining the closure of an element `x` as the infimum of all closed `y ≥ x` (assuming we're in a complete lattice). In particular, the emphasis is not on the **preorder of closed elements**, but on the **closedness predicate**. Morally, this is because a closure operator is the same as a Galois connection to the preorder of closed elements, but without mentioning the preorder of closed elements. That means that if we cared about the preorder of closed elements, we would not have wanted a closure operator in the first place, but a Galois connection.

All this said, here's what this PR does
* Add an `IsClosed` predicate to `ClosureOperator`. This redundant field is here to fix definitional equalities.
* Remove the set `closeds`. This is mostly replaced by the predicate, except in the map `toCloseds`, so we...
* ... Introduce an `abbrev Closeds` for the order of closed elements
* Rename `toClosed` to `toCloseds` to match.
* Rename the constructor `mk₃` to the more sexy `ofPred`. This is virtually the only useful constructor. Maybe it should even be taken as the definition, but I haven't gone that far.
* Rename the old `ofPred` to `ofCompletePred`. `ofCompletePred` is a specialisation of `ofPred` to the case where we are in a complete lattice.
* Remove a bunch of lemmas that are junk now that the definitional equality for `IsClosed` can be controlled.
* Golf existing uses of `ClosureOperator`. The diff here, I think, truly demonstrates that this PR is a step forward.

I am not extending the refactor to `LowerAdjoint` because:
* I am still not sure we want `LowerAdjoint` to even exist
* The rationale of preferring the closedness predicate over the preorder of closed elements is less strong here.
* I am lazy.

It is however easy to do the same refactor if wanted.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/Closed.lean [hunk 1]
```diff
@@ -104,18 +104,17 @@ theorem close_isClosed {X : C} (S : Sieve X) : J₁.IsClosed (J₁.close S) :=
...
-    apply le_antisymm
-    · intro Y f hf
-      rw [← J₁.covers_iff_mem_of_isClosed h]
-      apply hf
-    · apply J₁.le_close
-  · intro e
-    rw [← e]
-    apply J₁.close_isClosed
+theorem isClosed_iff_close_eq_self {X : C} (S : Sieve X) : J₁.IsClosed S ↔ J₁.close S = S :=
+  (J₁.closureOperator _).isClosed_iff
```
Comment (Vierkantor) [inline Mathlib/CategoryTheory/Sites/Closed.lean:117]: Nice work!

**Review by Vierkantor — COMMENTED** [review]
Do you think you can add a test case showing how this fixes the issue with `LowerAdjoint`?

LGTM otherwise, nice work!

**Comment by YaelDillies** [comment]
Sorry, what issue with `LowerAdjoint` am I supposed to have fixed? `LowerAdjoint` has the same disadvantage than `ClosureOperator` of being bundled, but without the benefit of only talking about one type. I'm claiming (but not acting on it here) that we could just delete `LowerAdjoint` and migrate the (very few) existing use cases to an unbundled `GaloisInsertion`.

**Comment by Vierkantor** [comment]
> Sorry, what issue with `LowerAdjoint` am I supposed to have fixed? `LowerAdjoint` has the same disadvantage than `ClosureOperator` of being bundled, but without the benefit of only talking about one type. I'm claiming (but not acting on it here) that we could just delete `LowerAdjoint` and migrate the (very few) existing use cases to an unbundled `GaloisInsertion`.

Oh, I must have misunderstood the Zulip discussion. Could you then explain a bit more concretely what wasn't possible before this refactor but is afterwards?

**Comment by YaelDillies** [comment]
I don't think it allows anything that wasn't possible before. However it clarifies the use cases of `ClosureOperator` vs `LowerAdjoint` vs `GaloisInsertion` and provides clearer API for talking about the preorder of closed elements.

Mostly, it seems it made @awainverse reconsider what they were about to do.

**Comment by Vierkantor** [comment]
I see, thanks for the explanation!

bors r+

**Pull request closed without merging**